query/functions/basic.feature

```diff
@@ -0,0 +1,153 @@
+    Given transaction commits
+    Given connection open write transaction for database: typedb
+    Given typeql write query
+      """
+      insert
+      $a isa person, has name "Abigail";
+      $b isa person, has name "Bernadette";
+      $c isa person, has name "Cliff";
+      $d isa person, has name "Damien";
+      $e isa person, has name "Eustace";
+      """
+    Given transaction commits
+
+    Given connection open read transaction for database: typedb
+    Given get answers of typeql read query
+      """
+      match
+       $x isa person, has name "Abigail";
+       $friend in friends-of($x);
+      """
+    Then answer size is: 5
+    Given get answers of typeql read query
+      """
+      match
+       $x isa person;
+       $friend in friends-of($x);
+      """
+    Then answer size is: 25
+
+
+  # TODO: Do we want to keep this? Taken from value-predicate
+  Scenario: attribute comparison can be used to classify concept pairs as predecessors and successors of each other
+    Given connection open schema transaction for database: typedb
+    Given typeql schema query
+      """
+      define
+
+      entity post,
+          plays reply-of:original,
+          plays reply-of:reply,
+          plays message-succession:predecessor,
+          plays message-succession:successor,
+          owns creation-date;
+
+      relation reply-of,
+          relates original,
+          relates reply;
+
+      relation message-succession,
+          relates predecessor,
+          relates successor;
+
+      attribute creation-date, value datetime;
+
+      fun message-successor-pairs() -> { post, post }:
+        match
+          (original:$p, reply:$s) isa reply-of;
+          $s has creation-date $d1;
+          $d1 < $d2;
+          (original:$p, reply:$r) isa reply-of;
+          $r has creation-date $d2;
+        return {$s, $r};
+      """
+    Given transaction commits
+
+    Given connection open write transaction for database: typedb
+    Given typeql write query
+      """
+      insert
+
+      $x isa post, has creation-date 2020-07-01;
+      $x1 isa post, has creation-date 2020-07-02;
+      $x2 isa post, has creation-date 2020-07-03;
+      $x3 isa post, has creation-date 2020-07-04;
+      $x4 isa post, has creation-date 2020-07-05;
+      $x5 isa post, has creation-date 2020-07-06;
+
+      (original:$x, reply:$x1) isa reply-of;
+      (original:$x, reply:$x2) isa reply-of;
+      (original:$x, reply:$x3) isa reply-of;
+      (original:$x, reply:$x4) isa reply-of;
+      (original:$x, reply:$x5) isa reply-of;
+      """
+    Given transaction commits
+
+    Given connection open read transaction for database: typedb
+    Given get answers of typeql read query
+      """
+      match $x1, $x2 in message-successor-pairs();
+      """
+    # the (n-1)th triangle number, where n is the number of replies to the first post
+    Then answer size is: 10
+
+
+  Scenario: A function can return a value derived from an expression
```

Comment:
Ok so i think we're missing some base cases right?

1. return a stream of non-tuple instances (-> `{person}`)
2. return a stream of non-tuple values (-> `{long}`)
3. return a n-tuple of instances + values (probably just mix in 1 test should be ok) (`{ x, y, z}`)
4. return a singular instance and value `-> person | -> long`
5. return tuples of instances and values `-> long, long, long`
6. Future (could write them now?) Optional variants of the above... ` -> { long?, person }` `-> person?` `-> long?` -> `long?, long, long`